vcf_annotation_tools/vep_annotation_reporter.py

```diff
@@ -0,0 +1,195 @@
+    )
+    parser.add_argument(
+        "-o", "--output-tsv",
+        help="Path to write the output report TSV file. If not provided, the output TSV will be written "
+            +"next to the input VCF with a .tsv file ending."
+    )
+    return parser
+
+def create_vcf_reader(args):
+    vcf_reader = vcfpy.Reader.from_path(args.input_vcf)
+    if 'CSQ' not in vcf_reader.header.info_ids():
+        vcf_reader.close()
+        raise Exception("ERROR: VCF {} is not VEP-annotated. Please annotate the VCF with VEP before running this tool.".format(args.input_vcf))
+    return vcf_reader
+
+def create_tsv_reader(input_filehandle):
+    tsv_reader = csv.DictReader(input_filehandle, delimiter = "\t")
+    for field in ['CHROM', 'POS', 'REF', 'ALT']:
+        if field not in tsv_reader.fieldnames:
+            raise Exception("ERROR: Input TSV {} doesn't contain required column '{}'.".format(input_filehandle.name, field))
+    return tsv_reader
+
+def parse_csq_header(vcf_reader):
+    format_pattern = re.compile('Format: (.*)')
+    return format_pattern.search(vcf_reader.header.get_info_field_info('CSQ').description).group(1).split('|')
+
+def parse_csq_entries(csq_entries, csq_fields):
+    transcripts = {}
+    for entry in csq_entries:
+        values = entry.split('|')
+        transcript = {}
+        for key, value in zip(csq_fields, values):
+            transcript[key] = value
+        if transcript['Allele'] not in transcripts.keys():
+            transcripts[transcript['Allele']] = []
+        transcripts[transcript['Allele']].append(transcript)
+    return transcripts
+
+def is_sv(entry):
+    return 'SVTYPE' in entry.INFO
+
+def resolve_alleles(entry, csq_alleles):
+    alleles = {}
+    if is_sv(entry):
+        for alt in entry.ALT:
+            alt = alt.serialize()
+            if len(alt) > len(entry.REF) and 'insertion' in csq_alleles:
+                alleles[alt] = 'insertion'
+            elif len(alt) < len(entry.REF) and 'deletion' in csq_alleles:
+                alleles[alt] = 'deletion'
+            elif len(csq_alleles) == 1:
+                alleles[alt] = list(csq_alleles)[0]
+    else:
+        for alt in entry.ALT:
+            alt = alt.serialize()
+            if len(alt) != len(entry.REF):
+                if alt[0:1] != entry.REF[0:1]:
+                    csq_allele = alt
+                elif alt[1:] == "":
+                    csq_allele = '-'
+                else:
+                    csq_allele = alt[1:]
+                alleles[alt] = csq_allele
+            else:
+                alleles[alt] = alt
+    return alleles
+
+def transcript_for_alt(transcripts, alt):
+    for transcript in transcripts[alt]:
+        if 'PICK' in transcript and transcript['PICK'] == '1':
+            return transcript
+    return transcripts[alt][0]
+
+def decode_hex(string):
+    hex_string = string.group(0).replace('%', '')
```

Comment:
This took me a minute to understand, because `string` is not a python `str` object (which lacks `group`).  It's a `Match` object from a regular expression.